rir/src/compiler/opt/types.cpp

```diff
@@ -40,33 +54,25 @@ void TypeInference::apply(RirCompiler&, ClosureVersion* function,
                     i->eachArg([&](Value* v) {
                         if (i->mayHaveEnv() && v == i->env())
                             return;
-                        if (auto arg = Instruction::Cast(v)) {
-                            if (types.count(arg))
-                                infered = infered | types.at(arg);
-                            else
-                                done = false;
-                        } else {
-                            infered = infered | v->type;
-                        }
+                        infered = infered | getType(v);
                     });
                     if (i->tag == Tag::Div && infered.isA(RType::integer)) {
                         infered = infered | RType::real;
                     }
                     break;
                 }
-                case Tag::Assume: {
-                    auto assumption = Assume::Cast(i);
-                    if (!assumption->assumeTrue)
-                        if (auto isO = IsObject::Cast(assumption->condition()))
-                            if (auto val =
-                                    Instruction::Cast(isO->arg<0>().val())) {
-                                if (types.count(val))
-                                    infered = types.at(val).notObject();
-                                else
-                                    infered = val->type.notObject();
-                            }
-                    break;
+                case Tag::CallSafeBuiltin: {
+                    auto c = CallSafeBuiltin::Cast(i);
+                    std::string name = getBuiltinName(getBuiltinNr(c->blt));
+                    if ("bitwiseXor" == name || "bitwiseShiftL" == name) {
```

Comment:
I'm always a bit wary when things like this are hardcoded... Isn't there a list of builtins?

Reply:
benchmark *cough*. just wanted to see it working :) I am not quite sure what the generic solution would be, since it highly depends on the actual builtin. also for each builtin you would basically need a matrix `t1 x t2 -> t` that tells you for every input combination, what the result type is.

Reply:
I'm still unhappy about hardcoding this, especially if it's just for the benchmark.

Are there any other simple cases we could add here?